File: haskell/2024/2024-001/part-002/main.hs
import System.IO
import Data.List (foldl')
import qualified Data.Map as Map

-- Main function to calculate similarity score
main :: IO ()
main = do
    -- File path
    let filePath = "input.txt"
    
    -- Read and parse the file
    content <- readFile filePath
    let (leftList, rightList) = parseFile content
    
    -- Create a frequency map for the right list
    let rightListCounts = createFrequencyMap rightList
    
    -- Calculate the similarity score
    let similarityScore = calculateSimilarityScore leftList rightListCounts
    
    -- Print the result
    putStrLn $ "Similarity Score: " ++ show similarityScore

-- Parse the file content into two lists
parseFile :: String -> ([Int], [Int])
parseFile content = 
    let linesList = lines content
        parseLine line = let [left, right] = map read (words line) in (left, right)
        (lefts, rights) = unzip $ map parseLine linesList
    in (lefts, rights)

-- Create a frequency map for a list of integers
createFrequencyMap :: [Int] -> Map.Map Int Int
createFrequencyMap = foldl' (\acc x -> Map.insertWith (+) x 1 acc) Map.empty

-- Calculate the similarity score
calculateSimilarityScore :: [Int] -> Map.Map Int Int -> Int
calculateSimilarityScore leftList rightListCounts =
    sum $ map (\left -> left * Map.findWithDefault 0 left rightListCounts) leftList

File: Swift/2024/2024-001/part-002/main.swift
import Foundation

// Function to calculate the similarity score
func calculateSimilarityScore(filePath: String) -> Int {
    // Read the file content
    guard let fileContent = try? String(contentsOfFile: filePath) else {
        print("Failed to read file at \(filePath)")
        return 0
    }
    
    // Split file content into lines
    let lines = fileContent.split(separator: "\n")
    
    // Initialize two lists
    var leftList: [Int] = []
    var rightList: [Int] = []
    
    // Parse the lines into two lists
    for line in lines {
        let parts = line.split(separator: " ").compactMap { Int($0) }
        if parts.count == 2 {
            leftList.append(parts[0])
            rightList.append(parts[1])
        }
    }
    
    // Create a dictionary to count occurrences in the right list
    var rightListCounts: [Int: Int] = [:]
    for number in rightList {
        rightListCounts[number, default: 0] += 1
    }
    
    // Calculate the similarity score
    let similarityScore = leftList.reduce(0) { (result, left) -> Int in
        let count = rightListCounts[left] ?? 0
        return result + (left * count)
    }
    
    return similarityScore
}

// Input file path
let filePath = "input.txt"

// Calculate and print the similarity score
let similarityScore = calculateSimilarityScore(filePath: filePath)
print("Similarity Score: \(similarityScore)")
